### PR TITLE
🎉 New source: Native Hubspot connector

### LINKED ISSUE #2114: Airbyte Native Hubspot connector
## Tell us about the problem you're trying to solve
We've recently received a few requests to support replicating more entities from our Hubspot connector. The connector is currently Singer-based and replicates the following entities: 
* "subscription_changes"
* "email_events"
* "forms"
* "workflows"
* "owners"
* "campaigns"
* "contact_lists"
* "contacts"
* "companies"
* "deals"
* "deal_pipelines"
* "engagements"
* "contacts_by_company"

We've received requests to support the following entities: 
* tickets
* products
* line_items
* quotes

Plus some extra fields in various entities (Contacts, Deals, Companies): 

property_country.updatedByUserId
property_country.versions
property_createdate.updatedByUserId
property_createdate.versions
property_d_u_n_s_number_c.updatedByUserId
property_d_u_n_s_number_c.versions
property_description.updatedByUserId
property_description.versions
property_dnb_business_name.updatedByUserId

## Describe the solution you’d like
We'd like to have a hubspot source which: 
* replicates the current set of entities 
* replicates the new requested entities and properties
* replicates data in an incremental fashion

Preferably this connector is written from scratch for the Airbyte protocol rather than forked from the singer hubspot tap. 

## Describe the alternative you’ve considered or used
Fork tap-hubspot, which is what our current connector is based on, to add support for these entities. The upside is that it's probably less work. The downside is that it'll carry Singer baggage in the code.

### LINKED ISSUE #2195: Source Hubspot: native client should support all streams in full sync mode
## Tell us about the problem you're trying to solve
To deliver an MVP we want to ensure that we can read from all streams

- [x] Add base API to perform authentication and GET requests
- [x] Add base stream to handle data fetching and pagination
- [x] Add a separate class for each stream

Dev estimate is 4-5h, because it requires reading a lot of documentation for API (for each stream)

### PR DESCRIPTION
**Pull request by eugene-kulak**
## What
closes #2195 
closes #2215 

## How
*Describe the solution*

## Pre-merge Checklist
- [ ] *Run integration tests*
- [ ] *Publish Docker images*

## Recommended reading order
1. `test.java`
1. `component.ts`
1. the rest

TODOs: 
- [x] Use interval
- [x] check if pagination is possible on contacts by company stream

### CODE REVIEW

**Review by yevhenii-ldv — APPROVED** [review]

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (sherifnada) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:233]: can you add context on what this is doing as a comment?

 I've looked at the API docs for a bit and it's not obvious to me what this is doing. it seems like it's adding the type of the property based on dynamically getting it from the Hubspot API? is this related to [this ticket](https://github.com/airbytehq/airbyte/issues/2219)? why wouldn't we know the type of the property ahead of time? 

This is bringing me to a different point: can you clarify as a docstring at the top of this class what the different APIs we are using are, and point to the API docs? This will enable me as the reader to know exactly which API docs to look at (Hubspot has multiple)

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (sherifnada) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:174]: `data_path` is a bit misleading as a name since it's a single key, not a path in the JSON. Can we use something like `data_field`?

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (sherifnada) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:179]: I'm adding all TODOs to a checklist in the PR description so we don't miss these before merging

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (sherifnada) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:256]: yield from?

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (sherifnada) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:200]: yield from?

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (sherifnada) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:307]: why does every class have `API` in the name? Shouldn't it just be the `API` class and then everything else is a stream?

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (sherifnada) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:320]: added to todo list

File: docs/integrations/sources/hubspot.md
Comment (sherifnada) [inline docs/integrations/sources/hubspot.md:13]: can you link to these in the docs? You don't have to link to each one individually, you can just link to one page at the top containing all of themnm

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (sherifnada) [inline airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:328]: is the point of this stream to be a join table between contacts and companies? can't we put this information in the contacts stream as a field?

**Review by sherifnada — COMMENTED** [review]
Looks good so far, I really like the breakdown of classes between `API`, StreamAPI, and CRMObjects 🚀

File: airbyte-config/init/src/main/resources/seed/source_definitions.yaml
Comment (eugene-kulak) [reply airbyte-config/init/src/main/resources/seed/source_definitions.yaml:81]: my intention was to add a completely new connector, then release it and drop the one that based on singer. But maybe accidentally renamed the old instead, not sure, need to check

File: airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py
Comment (eugene-kulak) [reply airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:233]: most objects in Hubspot have attributes located in the properties field, there are two types:
- default (built-in, depends on the version of API)
- custom (user-defined)
because the exact list and types are dynamic we retrieve them at runtime. Technically we don't need to know type of attribute here, but for building a schema object we will need it (#2219 )

File: airbyte-config/init/src/main/resources/seed/source_definitions.yaml
Comment (sherifnada) [reply airbyte-config/init/src/main/resources/seed/source_definitions.yaml:81]: Important thing is not to have 2 hubspot connectors in Airbyte at the same time. the user only cares about syncing data from hubspot, Singer vs. native is an implementation detail

**Review by yevhenii-ldv — APPROVED** [review]

File: airbyte-config/init/src/main/resources/seed/source_definitions.yaml
Comment (keu) [reply airbyte-config/init/src/main/resources/seed/source_definitions.yaml:81]: clear, I have removed the old connector

File: airbyte-integrations/connectors/source-hubspot/setup.py
Comment (sherifnada) [inline airbyte-integrations/connectors/source-hubspot/setup.py:36]: airbyte-python-test is not needed

File: docs/integrations/sources/hubspot.md
Comment (sherifnada) [inline docs/integrations/sources/hubspot.md:51]: As far as I can tell, this connector/airbyte would need to implement oauth for a client to be able to obtain oauth credentials. Since we don't do this yet, can you remove the oauth section from the docs and create an issue to support it once we have oauth? blocked on this issue: #768

**Review by sherifnada — APPROVED** [review]

File: docs/integrations/sources/hubspot.md
Comment (keu) [reply docs/integrations/sources/hubspot.md:51]: ok https://github.com/airbytehq/airbyte/issues/2449

**Pull request merged**
